From c92df3d93ad09f219e0728b39d40dc0a69d0760f Mon Sep 17 00:00:00 2001
From: Vijayakumar Durai <vijayakumar.durai1@vivint.com>
Date: Tue Feb 26 19:29:30 2019 +0530
Subject: [PATCH] rt2x00: do not increment sequence number while
 re-transmitting

Currently STA+AP re-transmitting the management frames with
incremented sequence number if hardware is assigning the sequence.

Fix is to assign the sequence number for Beacon by hardware
and for other Management frames software will assign the
sequence number

Signed-off-by: Vijayakumar Durai <vijayakumar.durai1@vivint.com>
---
 .../net/wireless/ralink/rt2x00/rt2x00queue.c  | 26 +++++++++----------
 1 file changed, 12 insertions(+), 14 deletions(-)

--- a/drivers/net/wireless/ralink/rt2x00/rt2x00queue.c
+++ b/drivers/net/wireless/ralink/rt2x00/rt2x00queue.c
@@ -207,22 +207,20 @@ static void rt2x00queue_create_tx_descri
 		 */
 		if (test_bit(CONFIG_QOS_DISABLED, &rt2x00dev->flags))
 			__clear_bit(ENTRY_TXD_GENERATE_SEQ, &txdesc->flags);
-		else
-			/* H/W will generate sequence number */
-			return;
+		else {
+			/*
+			 * rt2800 has a beacon problem(beacon is sent with same sequence
+			 * number more than once)with software generated sequence number.
+			 * So for beacons,hardware will generate sequence number and for
+			 * other Management frames software will generate sequence number.
+			 */
+			if (ieee80211_is_beacon(hdr->frame_control))
+				return;
+			else
+				__clear_bit(ENTRY_TXD_GENERATE_SEQ, &txdesc->flags);
+		}
 	}
 
-	/*
-	 * The hardware is not able to insert a sequence number. Assign a
-	 * software generated one here.
-	 *
-	 * This is wrong because beacons are not getting sequence
-	 * numbers assigned properly.
-	 *
-	 * A secondary problem exists for drivers that cannot toggle
-	 * sequence counting per-frame, since those will override the
-	 * sequence counter given by mac80211.
-	 */
 	if (test_bit(ENTRY_TXD_FIRST_FRAGMENT, &txdesc->flags))
 		seqno = atomic_add_return(0x10, &intf->seqno);
 	else
